cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp

```diff
@@ -0,0 +1,236 @@
+
+/*
+ * Copyright (c) 2019, Yung-Yu Chen <yyc@solvcon.net>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ * - Neither the name of the copyright holder nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <modmesh/buffer/SimpleArray.hpp>
+
+namespace modmesh
+{
+namespace python
+{
+
+template <typename T /* original type */, typename D /* for destination type */>
+struct TypeBroadcastImpl
+{
+    using slice_type = small_vector<int>;
+    using shape_type = typename SimpleArray<T>::shape_type;
+
+    static void broadcast(SimpleArray<T> & arr_out, std::vector<slice_type> const & slices, pybind11::array const & arr_in)
+    {
+        using out_type = typename std::remove_reference_t<decltype(arr_out[0])>;
+
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        auto * arr_new = reinterpret_cast<pybind11::array_t<D> const *>(&arr_in);
+
+        shape_type left_shape(arr_out.ndim());
+        for (size_t i = 0; i < arr_out.ndim(); i++)
+        {
+            slice_type const & slice = slices[i];
+            if ((slice[1] - slice[0]) % slice[2] == 0)
+            {
+                left_shape[i] = (slice[1] - slice[0]) / slice[2];
+            }
+            else
+            {
+                left_shape[i] = (slice[1] - slice[0]) / slice[2] + 1;
+            }
+        }
+
+        shape_type sidx_init(arr_out.ndim());
+
+        for (size_t i = 0; i < arr_out.ndim(); ++i)
+        {
+            sidx_init[i] = 0;
+        }
+
+        std::function<void(shape_type, int)> copy_idx;
```

Comment:
@yungyuc 
This code will cause an error. 
seems there is a symbol for lambda, but it cannot be used in another place?
```log
In file included from /home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:30:
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp: In instantiation of ‘static void modmesh::python::TypeBroadcastImpl<T, D>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; D = bool]’:
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:158:50:   required from ‘static void modmesh::python::TypeBroadcast<T>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:329:36:   required from ‘static void modmesh::python::WrapSimpleArray<T>::broadcast_array_using_slice(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:187:44:   required from ‘static void modmesh::python::WrapSimpleArray<T>::setitem_parser(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const pybind11::args&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:143:33:   required from ‘modmesh::python::WrapSimpleArray<T>::WrapSimpleArray(pybind11::module&, const char*, const char*) [with T = bool; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/common.hpp:266:29:   required from ‘static modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type& modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::commit(pybind11::module&, const char*, const char*) [with Wrapper = modmesh::python::WrapSimpleArray<bool>; Wrapped = modmesh::SimpleArray<bool>; Holder = std::unique_ptr<modmesh::SimpleArray<bool>, std::default_delete<modmesh::SimpleArray<bool> > >; WrappedBase = modmesh::SimpleArray<bool>; modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type = modmesh::python::WrapSimpleArray<bool>; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:377:76:   required from here
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, bool>::<lambda(modmesh::python::TypeBroadcastImpl<bool, bool>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, bool>::<lambda(modmesh::python::TypeBroadcastImpl<bool, bool>::shape_type, int)>::<arr_in capture>’ [-Werror=attributes]
   75 |         copy_idx = [&](shape_type sidx, int dim)
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   76 |         {
      |         ~           
   77 |             if (dim < 0)
      |             ~~~~~~~~~~~~
   78 |             {
      |             ~       
   79 |                 return;
      |                 ~~~~~~~
   80 |             }
      |             ~       
   81 | 
      |                     
   82 |             for (size_t i = 0; i < left_shape[dim]; ++i)
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   83 |             {
      |             ~       
   84 |                 sidx[dim] = i;
      |                 ~~~~~~~~~~~~~~
   85 | 
      |                     
   86 |                 size_t offset_in = 0;
      |                 ~~~~~~~~~~~~~~~~~~~~~
   87 |                 for (pybind11::ssize_t it = 0; it < arr_in.ndim(); ++it)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   88 |                 {
      |                 ~   
   89 |                     offset_in += arr_in.strides(it) / arr_in.itemsize() * sidx[it];
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   90 |                 }
      |                 ~   
   91 |                 const D * ptr_in = arr_new->data() + offset_in;
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   92 | 
      |                     
   93 |                 size_t offset_out = 0;
      |                 ~~~~~~~~~~~~~~~~~~~~~~
   94 |                 for (size_t it = 0; it < arr_out.ndim(); ++it)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   95 |                 {
      |                 ~   
   96 |                     auto step = slices[it][2];
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~
   97 |                     offset_out += arr_out.stride(it) * sidx[it] * step;
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   98 |                 }
      |                 ~   
   99 | 
      |                     
  100 |                 // NOLINTNEXTLINE(bugprone-signed-char-misuse, cert-str34-c)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  101 |                 arr_out.at(offset_out) = static_cast<out_type>(*ptr_in);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  102 |                 // recursion here
      |                 ~~~~~~~~~~~~~~~~~
  103 |                 copy_idx(sidx, dim - 1);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~
  104 |             }
      |             ~       
  105 |         };
      |         ~           
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, bool>::<lambda(modmesh::python::TypeBroadcastImpl<bool, bool>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, bool>::<lambda(modmesh::python::TypeBroadcastImpl<bool, bool>::shape_type, int)>::<arr_new capture>’ [-Werror=attributes]
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp: In instantiation of ‘static void modmesh::python::TypeBroadcastImpl<T, D>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; D = signed char]’:
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:162:52:   required from ‘static void modmesh::python::TypeBroadcast<T>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:329:36:   required from ‘static void modmesh::python::WrapSimpleArray<T>::broadcast_array_using_slice(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:187:44:   required from ‘static void modmesh::python::WrapSimpleArray<T>::setitem_parser(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const pybind11::args&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:143:33:   required from ‘modmesh::python::WrapSimpleArray<T>::WrapSimpleArray(pybind11::module&, const char*, const char*) [with T = bool; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/common.hpp:266:29:   required from ‘static modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type& modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::commit(pybind11::module&, const char*, const char*) [with Wrapper = modmesh::python::WrapSimpleArray<bool>; Wrapped = modmesh::SimpleArray<bool>; Holder = std::unique_ptr<modmesh::SimpleArray<bool>, std::default_delete<modmesh::SimpleArray<bool> > >; WrappedBase = modmesh::SimpleArray<bool>; modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type = modmesh::python::WrapSimpleArray<bool>; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:377:76:   required from here
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, signed char>::<lambda(modmesh::python::TypeBroadcastImpl<bool, signed char>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, signed char>::<lambda(modmesh::python::TypeBroadcastImpl<bool, signed char>::shape_type, int)>::<arr_in capture>’ [-Werror=attributes]
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, signed char>::<lambda(modmesh::python::TypeBroadcastImpl<bool, signed char>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, signed char>::<lambda(modmesh::python::TypeBroadcastImpl<bool, signed char>::shape_type, int)>::<arr_new capture>’ [-Werror=attributes]
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp: In instantiation of ‘static void modmesh::python::TypeBroadcastImpl<T, D>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; D = short int]’:
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:166:53:   required from ‘static void modmesh::python::TypeBroadcast<T>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:329:36:   required from ‘static void modmesh::python::WrapSimpleArray<T>::broadcast_array_using_slice(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:187:44:   required from ‘static void modmesh::python::WrapSimpleArray<T>::setitem_parser(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const pybind11::args&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:143:33:   required from ‘modmesh::python::WrapSimpleArray<T>::WrapSimpleArray(pybind11::module&, const char*, const char*) [with T = bool; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/common.hpp:266:29:   required from ‘static modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type& modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::commit(pybind11::module&, const char*, const char*) [with Wrapper = modmesh::python::WrapSimpleArray<bool>; Wrapped = modmesh::SimpleArray<bool>; Holder = std::unique_ptr<modmesh::SimpleArray<bool>, std::default_delete<modmesh::SimpleArray<bool> > >; WrappedBase = modmesh::SimpleArray<bool>; modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type = modmesh::python::WrapSimpleArray<bool>; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:377:76:   required from here
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, short int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, short int>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, short int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, short int>::shape_type, int)>::<arr_in capture>’ [-Werror=attributes]
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, short int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, short int>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, short int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, short int>::shape_type, int)>::<arr_new capture>’ [-Werror=attributes]
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp: In instantiation of ‘static void modmesh::python::TypeBroadcastImpl<T, D>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; D = int]’:
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:170:53:   required from ‘static void modmesh::python::TypeBroadcast<T>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:329:36:   required from ‘static void modmesh::python::WrapSimpleArray<T>::broadcast_array_using_slice(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:187:44:   required from ‘static void modmesh::python::WrapSimpleArray<T>::setitem_parser(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const pybind11::args&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:143:33:   required from ‘modmesh::python::WrapSimpleArray<T>::WrapSimpleArray(pybind11::module&, const char*, const char*) [with T = bool; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/common.hpp:266:29:   required from ‘static modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type& modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::commit(pybind11::module&, const char*, const char*) [with Wrapper = modmesh::python::WrapSimpleArray<bool>; Wrapped = modmesh::SimpleArray<bool>; Holder = std::unique_ptr<modmesh::SimpleArray<bool>, std::default_delete<modmesh::SimpleArray<bool> > >; WrappedBase = modmesh::SimpleArray<bool>; modmesh::python::WrapBase<Wrapper, Wrapped, Holder, WrappedBase>::wrapper_type = modmesh::python::WrapSimpleArray<bool>; pybind11::module = pybind11::module_]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:377:76:   required from here
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, int>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, int>::shape_type, int)>::<arr_in capture>’ [-Werror=attributes]
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:75:20: error: ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, int>::shape_type, int)>’ declared with greater visibility than the type of its field ‘modmesh::python::TypeBroadcastImpl<T, D>::broadcast<bool, int>::<lambda(modmesh::python::TypeBroadcastImpl<bool, int>::shape_type, int)>::<arr_new capture>’ [-Werror=attributes]
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp: In instantiation of ‘static void modmesh::python::TypeBroadcastImpl<T, D>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; D = long int]’:
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/TypeBroadcast.hpp:174:53:   required from ‘static void modmesh::python::TypeBroadcast<T>::broadcast(modmesh::SimpleArray<U>&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool]’
/home/mujin/modmesh/cpp/modmesh/python/wrapper/modmesh/wrap_SimpleArray.cpp:329:36:   required from ‘static void modmesh::python::WrapSimpleArray<T>::broadcast_array_using_slice(modmesh::python::WrapSimpleArray<T>::wrapped_type&, const std::vector<modmesh::small_vector<int> >&, const pybind11::array&) [with T = bool; modmesh::python::WrapSimpleArray<T>::wrapped_type = modmesh::SimpleArray<bool>]’

```

Reply:
It has something to do with https://gcc.gnu.org/wiki/Visibility .  And I wonder why don't you use `auto` for holding the lambda?  It does not seem that you are using the lambda outside the static member function function, and a `std::function` seems to be an overkill.

Reply:
I didn't use `auto` here because I had a deduction error. I don't know if it is able to address, so I just changed to use `std::function`

If I write as 
```
            auto copy_idx = [&](shape_type sidx, int dim) -> void
            {
            // ....
```

```
 error: use of ‘copy_idx’ before deduction of ‘auto’
 ```

Reply:
umm.. Related reading
[I thinks it's because I use it in recursive way, so the program cannot deduct it](https://stackoverflow.com/questions/2067988/recursive-lambda-functions-in-c11)

Reply:
Then why don't you make it a static member function or a standalone helper class template?

Reply:
> Then why don't you make it a static member function or a standalone helper class template?
sorry :(, I don't get it, could you explain what is the idea?

Reply:
do you mean making `copy_idx` a standalone class? Is it too complicated and unclear for the code?

Reply:
Yes.  A static member function, free function, or a helper class are worth a try.  I can only tell the complexity after seeing the code.